app/src/main/java/org/mozilla/fenix/perf/StartupTimeline.kt

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.perf
+
+import android.util.Log
+import androidx.annotation.UiThread
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import java.util.EnumMap
+import java.util.concurrent.TimeUnit
+
+private const val TAG = Performance.TAG
+
+/**
+ * Instruments points along the startup code path and logs these metrics.
+ *
+ * Currently, logged values are only guaranteed to be valid during cold startup to homescreen.
+ */
+@UiThread // to minimize overhead and simplify call patterns.
+object StartupTimeline {
+
+    enum class Method {
+        APP_ON_CREATE,
+        HOME_ACTIVITY_ON_CREATE,
+        HOME_ACTIVITY_ON_RESUME,
+        HOME_FRAGMENT_ON_CREATE,
+        HOME_FRAGMENT_ON_CREATE_VIEW,
+        HOME_FRAGMENT_ON_VIEW_CREATED,
+        HOME_FRAGMENT_ON_START,
+        HOME_FRAGMENT_ON_RESUME
+    }
+
+    private val methodStart = EnumMap<Method, Long>(Method::class.java)
+    private val methodEnd = EnumMap<Method, Long>(Method::class.java)
+
+    private var isColdStartDumpComplete = false
```

Comment:
We could use this flag that other way around:

1. Call it something like `isColdStart` and default to `true`
1. Only execute the main operations of `dumpInstrumentation()` when `isColdStart` is `true`.
1. Before completion of `dumpInstrumentation`, set `isColdStart` to `false`. 

In that case, when we are executing the dump, we know that the logged values are all non-null and that could simplify things? I think that this approach will work given that the lifetime of `object`s is tied to the lifetime of the application. Therefore, this flag will not get reset unless the entire application is restarted, exactly what we want. 

That's just an idea I had while thinking to myself.